### PR TITLE
[FIX] l10n_cl: Modify PDF to allow registering invoices in other currencies

### PR DESCRIPTION
**Pull request by Danisan**
Before this PR: 
PDF for invoices (after the new calculation methods added for l10n_cl_edi) where shown in different currencies than CLP,
which was incorrect.
With this fix, the PDF shows the amount in CLP but preserves each line amount in another currency.

Methods from l10n_cl_edi have been moved from l10n_cl_edi (enterprise) as part of #46593.
This includes _l10n_cl_get_amounts and _float_repr_float_round on the account move
model and the method _l10n_cl_get_line_amounts on the account move line model. 
They can therefore be removed from this commit.

The VAT section for the electronic ticket document is also being transferred in the same manner, and can thus be removed in the template here.

As part of the same refactor, methods _l10n_cl_get_invoice_totals_for_report and _l10n_cl_prices_and_taxes have been deprecated, and additionally, there is a new exclusive Qweb totals widget for Chilean invoices. This will allow showing in PDF reports, invoices that had been generated in foreign currencies or the UF index.

### CODE REVIEW

**Comment by Danisan** [comment]
@fvz-odoo this is the PR for the pdf printing in Odoo 16. There is also a PR associated to this one in enterprise.

cc: @jco-odoo

**Comment by Danisan** [comment]
@fvz-odoo @jco-odoo I found some problems in withholdings, net amounts, and coexistence with other l10ns. I am checking locally with deeper tests of cases and installing other l10ns

**Comment by Danisan** [comment]
@fvz-odoo and @jco-odoo I fixed the problem for withholdings and several other things for boletas.

File: addons/l10n_cl/views/report_invoice.xml [hunk 1]
```diff
@@ -208,9 +213,22 @@
             <div class="row">
                 <div name="stamp" class="col-4 text-center"/>
                 <div name="transferable-table" class="col-4"/>
-                <div name="transferable-legend" class="col-4 float-end"/>
+                <div name="transferable-legend" class="col-4 text-end">
+                    <t t-if="o.l10n_latam_document_type_id.code == '39' and o.country_code == 'CL'">
+                        The VAT tax of this boleta is: <span t-esc="o._l10n_cl_get_amounts()['vat_amount']" t-options="{'widget': 'monetary', 'display_currency': o.currency_id}"/>.
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:218]: total_amounts variable?

File: addons/l10n_cl/views/report_invoice.xml [hunk 2]
```diff
@@ -208,9 +213,22 @@
...
+                    </t>
+                </div>
             </div>
         </xpath>
+        <xpath expr="//span[@t-field='line.name']" position="replace">
+            <t t-set="line_amounts" t-value="line._l10n_cl_get_line_amounts()"/>
+            <t t-if="line_amounts.get('second_currency') and not o.l10n_latam_document_type_id._is_doc_type_export()">
+                <span t-esc="'%s (%s: %s @ %s)' % (line.name, o.currency_id.name, o._float_repr_float_round(line.price_unit, 2), o._float_repr_float_round(1/o.currency_id.rate, 2))" t-options="{'widget': 'text'}"/>
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:226]: line.price_unit should just need a float_repr

**Comment by Danisan** [comment]
@fvz-odoo @jco-odoo Please take a look at this. Tks!

File: addons/l10n_cl/views/report_invoice.xml [hunk 3]
```diff
@@ -44,20 +44,20 @@
                 <div name="right-upper-side" class="col-4">
                     <div class="row">
                         <div name="right-upper-side" class="col-12">
-                            <div class="row border border-dark">
+                            <div class="row border border-4 border-dark">
                                 <div class="col-12 text-center">
-                                    <h6 t-att-style="'color: %s;' % o.company_id.primary_color">
-                                        <strong t-att-style="'color: %s;' % o.company_id.primary_color">
+                                    <h6 style="color: black;">
```
Comment (Danisan) [inline addons/l10n_cl/views/report_invoice.xml:49]: Captura de pantalla 2023-10-02 a la(s) 09 18 18

File: addons/l10n_cl/views/report_invoice.xml [hunk 4]
```diff
@@ -229,4 +247,90 @@
...
 
-</odoo>
+    <template id="cl_document_tax_totals" inherit_id="account.document_tax_totals">
+
+        <t t-foreach="tax_totals['subtotals']" t-as="subtotal" position="before">
+            <t t-set="total_amounts" t-value="o._l10n_cl_get_amounts()"/>
+            <t t-set="withholdings" t-value="o._l10n_cl_get_withholdings()"/>
+        </t>
+        <t t-foreach="tax_totals['subtotals']" t-as="subtotal" position="replace">
+            <t t-if="o.country_code != 'CL' or o.l10n_latam_document_type_id._is_doc_type_electronic_ticket()">
```
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml:274]: Do you mean a different template for boletas?

File: addons/l10n_cl/views/report_invoice.xml [hunk 5]
```diff
@@ -229,4 +264,98 @@
...
+                              t-esc="total_amounts['subtotal_amount_exempt']"
+                              t-options="{'widget': 'monetary', 'display_currency': total_amounts['main_currency']}"
+                        />
+                    </td>
+                </tr>
+
+                <tr class="border-black o_subtotal" t-if="total_amounts['vat_amount'] != 0">
+                    <td class="border-black o_subtotal">
+                        <strong t-esc="'VAT %s' % total_amounts['vat_percent']"/><strong>%</strong>
+                        <strong>( Base: </strong>
```
Comment (Danisan) [inline addons/l10n_cl/views/report_invoice.xml:310]: With this changes, the footer is being shown like this:
Captura de pantalla 2023-10-02 a la(s) 09 30 13

Nevertheless: the example in the documentation gives this detail. It does not say anything about "base" amounts but it does not forbid it:
Captura de pantalla 2023-10-02 a la(s) 09 31 05

File: addons/l10n_cl/views/report_invoice.xml [hunk 6]
```diff
@@ -204,14 +223,30 @@
             <div t-attf-class="#{'col-6' if report_type != 'html' else 'col-sm-7 col-md-6'}"/>
         </xpath>
 
-        <xpath expr="//div[@id='total']" position="after">
+        <xpath expr="//div[@id='total']/div" position="inside">
+            <div name="stamp" class="col-11 text-center"/>
+        </xpath>
```
Comment (Danisan) [inline addons/l10n_cl/views/report_invoice.xml:228]: This ensure correct dimensions of the stamp:
Captura de pantalla 2023-10-02 a la(s) 09 33 58

Captura de pantalla 2023-10-02 a la(s) 09 40 17

File: addons/l10n_cl/views/report_invoice.xml [hunk 7]
```diff
@@ -156,12 +156,31 @@
...
 
+        <th name="th_price_unit" t-if="display_discount" t-attf-class="text-end {{ 'd-none d-md-table-cell' if report_type == 'html' else '' }}" position="after">
+            <th name="th_price_unit" t-if="display_discount" t-attf-class="text-end {{ 'd-none d-md-table-cell' if report_type == 'html' else '' }}">
+                <span>Disc.%</span>
+            </th>
+        </th>
+
+        <td t-if="display_discount" t-attf-class="text-end {{ 'd-none d-md-table-cell' if report_type == 'html' else '' }}" position="after">
+            <td t-if="display_discount" t-attf-class="text-end {{ 'd-none d-md-table-cell' if report_type == 'html' else '' }}">
+                <span class="text-nowrap" t-esc="line_amounts['total_discount']" t-options="{'widget': 'monetary', 'display_currency': line_amounts['main_currency']}"/>
```
Comment (Danisan) [inline addons/l10n_cl/views/report_invoice.xml:179]: comply with discount in amount and not in %

Captura de pantalla 2023-10-02 a la(s) 17 26 25

File: addons/l10n_cl/models/account_move.py [hunk 8]
```diff
@@ -161,6 +161,36 @@ def _l10n_cl_get_invoice_totals_for_report(self):
             tax_totals['formatted_amount_untaxed'] = formatLang(self.env, tax_totals['amount_untaxed'], currency_obj=self.currency_id)
             if tax_totals['subtotals']:
                 tax_totals['subtotals'][0]['formatted_amount'] = tax_totals['formatted_amount_untaxed']
+        values = self._l10n_cl_get_amounts()
```
Comment (Danisan) [inline addons/l10n_cl/models/account_move.py:164]: In this last commit, instead of modifying the totals section with a new inherited template, the idea is to use the pre-existing method `_l10n_cl_get_invoice_totals_for_report()` and add the missing information to that dictionary. The final result is like this:

Captura de pantalla 2023-10-03 a la(s) 13 03 10

**Review by Danisan — COMMENTED** [review]
@jco-odoo and @fvz-odoo see the last proposals

**Comment by fvz-odoo** [comment]
Hello @jco-odoo 

I just tested this and functionally it is working correctly. You can go ahead and review... (attached is an example of one of the PDFs I tested).

Cheers!

Fernando
[FAC 000002.pdf](https://github.com/odoo/odoo/files/12807502/FAC.000002.pdf)


Thanks for the PR @Danisan 

@ren-odoo

File: addons/l10n_cl/views/report_invoice.xml (hunk 1 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: Well, this is fixed, since now we call the method `_l10n_cl_get_amounts()` from `_l10n_cl_get_invoice_totals_for_report`

File: addons/l10n_cl/views/report_invoice.xml (hunk 2 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: we made the same thing for the lines in this commit 0659f4d we call `_l10n_cl_get_line_amounts()`from inside the method `_l10n_cl_prices_and_taxes()`

**Comment by Danisan** [comment]
@ren-odoo @fvz-odoo and @jco-odoo: I made an additional change where I call the method for the line calculation from an already existing method for that.

File: addons/l10n_cl/models/account_move.py [hunk 9]
```diff
@@ -160,6 +161,36 @@ def _l10n_cl_get_invoice_totals_for_report(self):
...
+        withholdings = self._l10n_cl_get_withholdings()
+        if self.currency_id != self.env.ref('base.CLP'):
+            tax_totals['formatted_amount_total'] = formatLang(
+                self.env, values['total_amount'], currency_obj=self.env.ref('base.CLP'))
+            tax_totals['subtotals'][0]['amount'] = values['subtotal_amount_taxable']
+            tax_totals['subtotals'][0]['formatted_amount'] = formatLang(
+                self.env, values['subtotal_amount_taxable'], currency_obj=self.env.ref('base.CLP'))
+        list_of_tax_groups = []
+        for groups_by_subtotal in tax_totals['groups_by_subtotal']['Base imponible']:
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move.py:173]: if _('Untaxed Amount') in totals['groups_by_subtotal']:

File: addons/l10n_cl/models/account_move.py [hunk 10]
```diff
@@ -160,6 +161,36 @@ def _l10n_cl_get_invoice_totals_for_report(self):
...
+            if groups_by_subtotal['tax_group_name'] != 'IVA 19%':
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_cl/models/account_move.py:174]: tax.group.search(l10n_cl_sii_code = 14)
groups_by_subtotal['tax_group_id'] in tax_group_iva

File: addons/l10n_cl/models/account_move.py [hunk 11]
```diff
@@ -160,6 +161,36 @@ def _l10n_cl_get_invoice_totals_for_report(self):
...
+                'tax_group_name': withholding['tax_name'],
+                'formatted_tax_group_amount': formatLang(self.env, withholding['tax_amount'], currency_obj=self.env.ref('base.CLP')),
+                'formatted_tax_group_base_amount': formatLang(self.env, withholding['tax_base'], currency_obj=self.env.ref('base.CLP'))
+            })
+        tax_totals['groups_by_subtotal']['Base imponible'] = list_of_tax_groups
+        if values['subtotal_amount_exempt']:
+            tax_totals['subtotals'].append({
+                'amount': values['subtotal_amount_exempt'],
+                'formatted_amount': formatLang(self.env, values['subtotal_amount_exempt'], currency_obj=self.env.ref('base.CLP')),
+                'name': 'Monto Exento'})
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move.py:192]:
```suggestion
                'name': _("Exempt Amount")})
```

File: addons/l10n_cl/models/account_move.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py]: fixed here: e2422049e

File: addons/l10n_cl/models/account_move.py (hunk 11 above)
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py]: I made this commit, but at least locally, it does not work for me.

File: addons/l10n_cl/models/account_move.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py]: this did not work

Comment (Danisan) [reply addons/l10n_cl/models/account_move.py]: cff2343c but didn't work

File: addons/l10n_cl/models/account_move.py (hunk 11 above)
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py]: b7944a0c

**Comment by Danisan** [comment]
@jco-odoo check the keys based in translations... I made a provisional commit for this, but in local did not get to work, so it would be nice that it could be tested in runbot. please
cc @fvz-odoo

**Comment by fvz-odoo** [comment]
Hello @Danisan 

Thanks for the changes, @jco-odoo this looks good to me now. 

Let us know if anything.

Fernando

@ren-odoo

File: addons/l10n_cl/models/account_move_line.py [hunk 12]
```diff
@@ -9,23 +11,102 @@ class AccountMoveLine(models.Model):
...
+        values = self._l10n_cl_get_line_amounts()
+        values['line_description'] = '%s (%s: %s @ %s)' % (
+            self.name,
+            values['second_currency']['currency_name'],
+            float_repr(values['second_currency']['price'], values['second_currency']['round_currency']),
+            self.move_id._float_repr_float_round(values['second_currency']['conversion_rate'], values['second_currency']['round_currency']),
+        ) if values.get('second_currency') and not self.l10n_latam_document_type_id._is_doc_type_export() else self.name
         return {
-            'price_unit': price_unit,
+            'values': values,
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move_line.py:35]:
```suggestion
            'line_amounts': values,
```

**Review by jco-odoo — COMMENTED** [review]
Maybe we should not change _l10n_cl_prices_and_taxes, remove that in master and just rely on the line amounts method.

File: addons/l10n_cl/models/account_move_line.py (hunk 12 above)
Comment (Danisan) [reply addons/l10n_cl/models/account_move_line.py]: fixed here: 04d1d9d2

File: addons/l10n_cl/views/report_invoice.xml [hunk 13]
```diff
@@ -154,14 +154,33 @@
 
         <t t-set="current_subtotal" t-value="current_subtotal + line.price_subtotal" position="before">
             <t t-set="l10n_cl_values" t-value="line._l10n_cl_prices_and_taxes()"/>
+            <t t-set="line_amounts" t-value="l10n_cl_values['values']"/>
         </t>
 
+        <xpath expr="//span[@t-field='line.price_unit']" position="before">
+            <t t-set="all_taxes" t-value="'all_taxes'"/>
```
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: removed the `all_taxes` from here, it in other line already

**Comment by Danisan** [comment]
@fvz-odoo @jco-odoo this is ready. Please check / test. Thanks!

**Comment by Danisan** [comment]
@jco-odoo @fvz-odoo I removed the unneeded diffs in the latest commit. Would be nice if you could check again.

**Comment by fvz-odoo** [comment]
Hello @jco-odoo 

This one is ready as well, I have tested and it is working correctly.

Let me know if anything.

Fernando

@Danisan @ren-odoo

File: addons/l10n_cl/models/account_move_line.py [hunk 14]
```diff
@@ -9,9 +10,14 @@ class AccountMoveLine(models.Model):
...
         self.ensure_one()
+        if self.display_type != 'product':
+            return {
+                'line_amounts': {'line_description': self.name},
+                'price_subtotal': 0,
+            }
         invoice = self.move_id
         included_taxes = self.tax_ids.filtered(lambda x: x.l10n_cl_sii_code == 14) if self.move_id._l10n_cl_include_sii() else self.tax_ids
-        if not included_taxes:
+        if included_taxes:
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move_line.py:20]: Why do we change the logic if a bit further we override the logic by the _l10n_cl_get_line_amounts?  Maybe we should just leave this method alone and indicate it as "deprecated, to be removed in master".  
Maybe it was done here to have the line_description, but then we could add it in the _get_line_amounts as well as we need to add the line_description to the qweb anyways.

File: addons/l10n_cl/models/account_move.py [hunk 15]
```diff
@@ -169,5 +206,104 @@ def _l10n_cl_include_sii(self):
...
+        tax = [{'tax_code': line.tax_line_id.l10n_cl_sii_code,
+                'tax_name': line.tax_line_id.name,
+                'tax_base': abs(sum(self.invoice_line_ids.filtered(
+                    lambda x: line.tax_line_id.l10n_cl_sii_code in x.tax_ids.mapped('l10n_cl_sii_code')).mapped(
+                    'balance'))),
+                'tax_percent': abs(line.tax_line_id.amount),
+                'tax_amount_currency': self.currency_id.round(abs(line.amount_currency)),
+                'tax_amount': self.currency_id.round(abs(line.balance))} for line in self.line_ids.filtered(
+            lambda x: x.tax_group_id.id in [
+                self.env.ref('l10n_cl.tax_group_ila').id, self.env.ref('l10n_cl.tax_group_retenciones').id])]
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move.py:305]: there is no way to try to use the tax codes themselves instead of the tax group? (if someone does custom tax groups, nothing will work anymore)

**Review by jco-odoo — COMMENTED** [review]
I am a bit surprised we do not need the amounts in the invoice currency. (except on the price unit).  

Still thinking if I could suggest some code myself.

**Comment by jco-odoo** [comment]
@Danisan Is there a reason as well to not show the taxes on the invoice line in the PDF?

**Comment by jco-odoo** [comment]
@Danisan suggestion here: https://github.com/odoo-dev/odoo/commit/2d3e90cadd3b6e31ead0b847e21f283b711b0618

File: addons/l10n_cl/views/report_invoice.xml [hunk 16]
```diff
@@ -154,14 +154,32 @@
...
+        <xpath expr="//span[@t-field='line.price_unit']" position="before">
+            <t t-if="'second_currency' in line_amounts" t-set="line_second_currency_round" t-value="line_amounts['second_currency']['round_currency']"/>
+        </xpath>
+
         <xpath expr="//span[@t-field='line.price_unit']" position="attributes">
             <attribute name="t-field"></attribute>
-            <attribute name="t-esc">line.price_unit</attribute>
-            <attribute name="t-options">{"widget": "monetary", "display_currency": o.currency_id}</attribute>
+            <attribute t-if="not o.l10n_latam_document_type_id._is_doc_type_electronic_ticket()" name="t-esc">line_amounts['price_item_document']</attribute>
+            <attribute name="t-options">{"widget": "float", "precision": 2}</attribute>
```
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml:167]: Hello, @jco-odoo so, if we put at it is now (widget float) it's ok, right? remember that the reason we have price_item_document in the dict, is because in some cases we use price_unit and in other cases we use the value we get from balance divided by the quantity (discounts, etc).

File: addons/l10n_cl/models/account_move.py [hunk 17]
```diff
@@ -160,6 +161,42 @@ def _l10n_cl_get_invoice_totals_for_report(self):
...
+        for groups_by_subtotal in tax_totals['groups_by_subtotal'][_('Untaxed Amount')]:
+            group_id = tax_lines.tax_line_id.filtered(lambda x: x.l10n_cl_sii_code == 14).tax_group_id.id
+            if groups_by_subtotal['tax_group_id'] != group_id:
+                continue
+            groups_by_subtotal['formatted_tax_group_amount'] = formatLang(
+                self.env, values['vat_amount'], currency_obj=values['main_currency'])
+            groups_by_subtotal['formatted_tax_group_base_amount'] = formatLang(
+                self.env, values['subtotal_amount_taxable'], currency_obj=values['main_currency'])
+            list_of_tax_groups.append(groups_by_subtotal)
+        for withholding in withholdings:
```
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py:184]: because if we don't, we will have amounts in other currency instead of CLP.

**Comment by Danisan** [comment]
> @Danisan Is there a reason as well to not show the taxes on the invoice line in the PDF?

most of the customers ask us to remove the taxes from the lines since they are put in the subtotal/total sections. This saves space (columns) and the customers usually ask that. Besides, understanding that we have to put the discounts in amount (as you can see in the pdf doc specs from the sii) is less space occupied by numbers.

File: addons/l10n_cl/models/account_move.py (hunk 15 above)
Comment (Danisan) [reply addons/l10n_cl/models/account_move.py:305]: You are right. Nevertheless, it was made that way to have some kind of consistency with other l10ns. But perhaps we can do some work around this.

File: addons/l10n_cl/views/report_invoice.xml (hunk 16 above)
[the remark this answers is not in the document]
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml:167]: Besides, as here we use CLP or even other currencies, monetary should not apply, so I think it's OK at it is now... @jco-odoo and we already removed the monetary...

File: addons/l10n_cl/models/account_move_line.py (hunk 14 above)
Comment (Danisan) [reply addons/l10n_cl/models/account_move_line.py:20]: @jco-odoo This solution impacts 3 versions:
16.0 -> Leave at it is? (we talked that in 16 we kept it)
17.0 -> Define as deprecated the pdf methods. (?)
master -> remove pdf method completely (?)

**Comment by jco-odoo** [comment]
> @Danisan suggestion here: [odoo-dev@2d3e90c](https://github.com/odoo-dev/odoo/commit/2d3e90cadd3b6e31ead0b847e21f283b711b0618)

That suggestion seems like a good improvement already.  Could you apply it?

File: addons/l10n_cl/models/account_move.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_cl/models/account_move.py:184]: @Danisan We can also change the amounts if we update the existing elements.

**Comment by Danisan** [comment]
@jco-odoo I added the changes you suggested. I will make some tests (and perhaps changes over this). Do you mean that if these changes worked, we would keep this almost as seen now?
And we would do the rest of the changes (removing the method _l10n_cl_prices_and_taxes()) for the further branches?

**Comment by Danisan** [comment]
@jco-odoo I've tested this, and it worked. Nevertheless, I kept the lines:

```        
if self.display_type != 'product':
      return {
          'line_amounts': {'line_description': self.name},
          'price_subtotal': 0,
      }
 ```

Otherwise, if you've got sections or notes, you will have an exception.

**Comment by fvz-odoo** [comment]
Hello @Danisan 

I tested this and functionally it is working correctly. @jco-odoo 

Cheers!

Fernando

**Comment by jco-odoo** [comment]
@Danisan Yes, we can do the removal in master for the _get_prices...  

But we discussed that we would check if we could do something better for the tax totals as well as these could be done in a totally custom way.

**Comment by fvz-odoo** [comment]
Hello @Danisan 

Thanks for the changes, this works great. This is ready @jco-odoo 

Cheers!

Fernando

File: addons/l10n_cl/views/report_invoice.xml [hunk 18]
```diff
@@ -229,4 +250,69 @@
...
+                </t>
+                <t t-if="subtotal_amounts['vat_amount'] != 0.0">
+                    <tr>
+                        <td t-esc="'IVA %s' % subtotal_amounts['vat_percent']"/>
+                        <td class="text-end" t-esc="subtotal_amounts['vat_amount']"
+                            t-options="{'widget': 'monetary', 'display_currency': subtotal_amounts['main_currency']}"/></tr>
+                </t>
+                <t t-foreach="withholdings" t-as="wh">
+                    <tr>
+                        <td t-esc="'Ret %s sobre %s' % (wh['tax_name'], o._format_lang_totals(wh['tax_base'], subtotal_amounts['main_currency']))"/>
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:290]: Should be in English?

**Review by jco-odoo — COMMENTED** [review]
I did not see you did everything in the qweb from the basic get_amounts method...  It seems nice.

File: addons/l10n_cl/views/report_invoice.xml [hunk 19]
```diff
@@ -229,4 +250,69 @@
...
+
+                    <t t-set="subtotal_to_show" t-value="subtotal['name']"/>
+                    <t t-call="account.tax_groups_totals"/>
+                </t>
+            </t>
+            <t t-else="">
+                <t t-set="subtotal_amounts" t-value="o._l10n_cl_get_amounts()"/>
+                <t t-set="withholdings" t-value="o._l10n_cl_get_withholdings()"/>
+                <t t-if="subtotal_amounts['subtotal_amount_taxable'] != 0.0">
+                    <tr class="border-black is-subtotal"><td><strong>Amount Untaxed</strong></td><td class="text-end oe_subtotal_footer_separator" t-esc="subtotal_amounts['subtotal_amount_taxable']" t-options="{'widget': 'monetary', 'display_currency': subtotal_amounts['main_currency']}"/></tr>
```
Comment (fvz-odoo) [inline addons/l10n_cl/views/report_invoice.xml:273]: Change this "Amount Untaxed" to "Net Amount".

File: addons/l10n_cl/views/report_invoice.xml [hunk 20]
```diff
@@ -229,4 +250,69 @@
...
+                </t>
+                <t t-if="subtotal_amounts['subtotal_amount_exempt'] != 0.0">
+                    <tr class="border-black is-subtotal">
+                        <td><strong>Amount Exempt</strong></td>
```
Comment (fvz-odoo) [inline addons/l10n_cl/views/report_invoice.xml:277]: Change this "Amount Exempt" to "Exempt Amount"

File: addons/l10n_cl/views/report_invoice.xml [hunk 21]
```diff
@@ -229,4 +250,69 @@
...
+                        <td class="text-end oe_subtotal_footer_separator" t-esc="subtotal_amounts['subtotal_amount_exempt']"
+                            t-options="{'widget': 'monetary', 'display_currency': subtotal_amounts['main_currency']}"/>
+                    </tr>
+                </t>
+                <t t-if="subtotal_amounts['vat_amount'] != 0.0">
+                    <tr>
+                        <td t-esc="'IVA %s' % subtotal_amounts['vat_percent']"/>
```
Comment (fvz-odoo) [inline addons/l10n_cl/views/report_invoice.xml:284]: Change IVA to VAT

File: addons/l10n_cl/views/report_invoice.xml [hunk 22]
```diff
@@ -229,4 +250,69 @@
...
+                    <tr>
+                        <td t-esc="'Ret %s sobre %s' % (wh['tax_name'], o._format_lang_totals(wh['tax_base'], subtotal_amounts['main_currency']))"/>
+                        <td class="text-end" t-esc="wh['tax_amount']" t-options="{'widget': 'monetary', 'display_currency': subtotal_amounts['main_currency']}"/></tr>
+                </t>
+            </t>
+        </t>
+        <tr class="border-black o_total" position="replace">
+            <tr class="border-black o_total" position="replace">
+                <t t-if="o.country_code != 'CL'">
+                    <tr class="border-black o_total">
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:298]: $0 ?

File: addons/l10n_cl/views/report_invoice.xml [hunk 23]
```diff
@@ -229,4 +250,69 @@
...
+                </t>
+                <t t-foreach="withholdings" t-as="wh">
+                    <tr>
+                        <td t-esc="'Ret %s sobre %s' % (wh['tax_name'], o._format_lang_totals(wh['tax_base'], subtotal_amounts['main_currency']))"/>
+                        <td class="text-end" t-esc="wh['tax_amount']" t-options="{'widget': 'monetary', 'display_currency': subtotal_amounts['main_currency']}"/></tr>
+                </t>
+            </t>
+        </t>
+        <tr class="border-black o_total" position="replace">
+            <tr class="border-black o_total" position="replace">
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:296]: double replace?

File: addons/l10n_cl/views/report_invoice.xml [hunk 24]
```diff
@@ -229,4 +250,69 @@
         </xpath>
     </template>
 
-</odoo>
+    <template id="tax_totals_widget" inherit_id="account.document_tax_totals">
+        <t t-foreach="tax_totals['subtotals']" t-as="subtotal" position="replace">
+            <t t-if="o.country_code != 'CL'">
+                <t t-foreach="tax_totals['subtotals']" t-as="subtotal">
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:256]: $0?

File: addons/l10n_cl/views/report_invoice.xml [hunk 25]
```diff
@@ -229,4 +250,69 @@
...
+                        <td><strong t-esc="subtotal['name']"/></td>
+
+                        <td class="text-end">
+                            <span t-att-class="oe_subtotal_footer_separator" t-esc="subtotal['formatted_amount']"/>
+                        </td>
+                    </tr>
+
+                    <t t-set="subtotal_to_show" t-value="subtotal['name']"/>
+                    <t t-call="account.tax_groups_totals"/>
+                </t>
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:267]:
```suggestion
                $0
```

File: addons/l10n_cl/models/res_partner.py [hunk 26]
```diff
@@ -37,8 +37,7 @@ def _format_vat_cl(self, values):
         )
         partner_country_is_chile = country.code == "CL" or identification_type.country_id.code == "CL"
         if partner_country_is_chile and \
-                values.get('l10n_latam_identification_type_id') in identification_types and values.get('vat') and\
-                stdnum.util.get_cc_module('cl', 'vat').is_valid(values['vat']):
+                values.get('l10n_latam_identification_type_id') in identification_types and values.get('vat'):
```
Comment (jco-odoo) [inline addons/l10n_cl/models/res_partner.py:40]: unnecessary diff?

File: addons/l10n_cl/models/account_move.py [hunk 27]
```diff
@@ -169,5 +173,104 @@ def _l10n_cl_include_sii(self):
 
     def _is_manual_document_number(self):
         if self.journal_id.company_id.country_id.code == 'CL':
-            return self.journal_id.type == 'purchase' and not self.l10n_latam_document_type_id._is_doc_type_vendor()
+            return (self.journal_id.type == 'purchase' and not self.l10n_latam_document_type_id._is_doc_type_vendor())
```
Comment (jco-odoo) [inline addons/l10n_cl/models/account_move.py:176]: unnecessary change

File: addons/l10n_cl/views/report_invoice.xml [hunk 28]
```diff
@@ -229,4 +250,69 @@
         </xpath>
     </template>
 
-</odoo>
+    <template id="tax_totals_widget" inherit_id="account.document_tax_totals">
```
Comment (jco-odoo) [inline addons/l10n_cl/views/report_invoice.xml:253]:
```suggestion
    <template id="tax_totals_widget" inherit_id="account.document_tax_totals" primary="True">
```

File: addons/l10n_cl/views/report_invoice.xml [hunk 29]
```diff
@@ -204,14 +213,26 @@
...
+        <xpath expr="//div[@id='total']/div" position="inside">
+            <div name="stamp" class="col-11 text-center"/>
+        </xpath>
+
+        <xpath expr="//div[@id='total']//div//table" position="after">
             <div class="row">
-                <div name="stamp" class="col-4 text-center"/>
-                <div name="transferable-table" class="col-4"/>
-                <div name="transferable-legend" class="col-4 float-end"/>
+                <div class="col-12 text-end" t-if="o.l10n_latam_document_type_id.code == '39' and o.country_code == 'CL'" name="vat_boleta">
```
Comment (fvz-odoo) [inline addons/l10n_cl/views/report_invoice.xml:222]: Remove the o.country_code

File: addons/l10n_cl/views/report_invoice.xml (hunk 19 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: fixed 254506c926

File: addons/l10n_cl/views/report_invoice.xml (hunk 20 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: Fixed 254506c926

File: addons/l10n_cl/views/report_invoice.xml (hunk 22 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: removed since instead of that we don't use $0 because theoretically with don't need to check for country code here 254506c926

File: addons/l10n_cl/views/report_invoice.xml (hunk 18 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: fixed 254506c926

File: addons/l10n_cl/views/report_invoice.xml (hunk 21 above)
Comment (Danisan) [reply addons/l10n_cl/views/report_invoice.xml]: fixed b71616f9701

**Comment by Danisan** [comment]
@fvz-odoo the small changes are ready. I will work in translations. Meanwhile you can test

**Comment by fvz-odoo** [comment]
Hello @Danisan 

As @jco-odoo said, it gives an error in the sales template. The invoice PDF works correctly, but when trying to print the sales order template I get this traceback: 
Screenshot 2023-12-21 at 12 41 36 PM

Could you please check what's causing the error?

Thanks

Fernando

@ren-odoo @vbe-odoo

**Comment by Danisan** [comment]
@fvz-odoo @jco-odoo the last two fixes address the issue. So the tax widget is applied only to 'account.move' model.

**Comment by jco-odoo** [comment]
@robodoo r+
